core/node/l1_recovery/src/l1_fetcher/l1_fetcher.rs

```diff
@@ -0,0 +1,1083 @@
+    type Key<'a> = H256;
+
+    fn encode_key(key: Self::Key<'_>) -> String {
+        format!("commit_block_{}.bin", hex::encode(key))
+    }
+
+    serialize_using_bincode!();
+}
+
+#[allow(clippy::enum_variant_names)]
+#[derive(Error, Debug)]
+pub enum L1FetchError {
+    #[error("get logs failed")]
+    GetLogs,
+
+    #[error("get tx failed")]
+    GetTx,
+
+    #[error("get end block number failed")]
+    GetEndBlockNumber,
+}
+
+#[derive(Debug, Clone)]
+pub enum ProtocolVersioning {
+    OnlyV3,
+    AllVersions {
+        v2_start_batch_number: u64,
+        v3_start_batch_number: u64,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub struct L1FetcherConfig {
+    pub block_step: u64,
+
+    pub diamond_proxy_addr: Address,
+
+    pub versioning: ProtocolVersioning,
+}
+
+pub struct L1Fetcher {
+    eth_client: Box<DynClient<L1>>,
+    config: L1FetcherConfig,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum RollupEventType {
+    Commit,
+    #[allow(unused)]
+    Prove,
+    Execute,
+    NewPriorityTx,
+}
+
+pub enum RollupEventsFilter {
+    L1BatchNumber(L1BatchNumber),
+    PriorityOpId(PriorityOpId),
+    None,
+}
+
+impl L1Fetcher {
+    pub fn new(config: L1FetcherConfig, eth_client: Box<DynClient<L1>>) -> Result<Self> {
+        tracing::info!("Initialized L1 fetcher with config: {config:?}");
+        Ok(L1Fetcher { eth_client, config })
+    }
+
+    fn hyperchain_contract() -> Contract {
+        hyperchain_contract()
+    }
+
+    fn commit_functions() -> Result<Vec<Function>> {
+        let contract = Self::hyperchain_contract();
+        Ok(vec![
+            contract.functions_by_name("commitBatches").unwrap()[0].clone(),
+            contract
+                .functions_by_name("commitBatchesSharedBridge")
+                .unwrap()[0]
+                .clone(),
+        ])
+    }
+
+    fn rollup_event_by_type(event_type: RollupEventType) -> Result<Event> {
+        let event_name = match event_type {
+            RollupEventType::Commit => "BlockCommit",
+            RollupEventType::Prove => "BlocksVerification",
+            RollupEventType::Execute => "BlockExecution",
+            RollupEventType::NewPriorityTx => "NewPriorityRequest",
+        };
+        Ok(Self::hyperchain_contract().events_by_name(event_name)?[0].clone())
+    }
+
+    fn extract_l1_batch_number_from_rollup_event_log(
+        event_type: RollupEventType,
+        log: &Log,
+    ) -> L1BatchNumber {
+        let topic_index = match event_type {
+            RollupEventType::Commit => 1,
+            RollupEventType::Prove => 2,
+            RollupEventType::Execute => 1,
+            _ => panic!("{event_type:?} event doesn't have l1_batch_number"),
+        };
+        L1BatchNumber(log.topics[topic_index].to_low_u64_be() as u32)
+    }
+
+    fn extract_priority_op_id_from_rollup_event_log(
+        event_type: RollupEventType,
+        log: &Log,
+    ) -> PriorityOpId {
+        assert_eq!(event_type, RollupEventType::NewPriorityTx);
```

Comment:
why do you need the event_type argument in the first place, if you are just asserting the value?